external-crates/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -97,11 +99,164 @@ impl<'a, 'b> TypeView for TypeWithLoader<'a, 'b> {
     }
 }
 
+pub trait InterpreterInterface {
+    fn get_stack_len(&self) -> usize;
+    fn get_internal_state(&self) -> ExecutionState;
+    fn set_location(&self, err: PartialVMError) -> VMError;
+    fn maybe_core_dump(&self, err: VMError, frame: &Frame) -> VMError;
+}
+
+impl InterpreterInterface for Interpreter {
+    fn get_stack_len(&self) -> usize {
+        self.get_stack_len()
+    }
+    fn get_internal_state(&self) -> ExecutionState {
+        self.get_internal_state()
+    }
+
+    fn set_location(&self, err: PartialVMError) -> VMError {
+        self.set_location(err)
+    }
+
+    fn maybe_core_dump(&self, err: VMError, frame: &Frame) -> VMError {
+        self.maybe_core_dump(err, frame)
+    }
+}
+
 impl Interpreter {
     /// Limits imposed at runtime
     pub fn runtime_limits_config(&self) -> &VMRuntimeLimitsConfig {
         &self.runtime_limits_config
     }
+
+    pub fn handle_plugin_error(result: VMResult<()>, is_critical: bool) -> VMResult<()> {
+        match result {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                if is_critical {
+                    return Err(err);
+                } else {
+                    error!("Plugin error: {:?}", err);
+                }
+                Ok(())
+            }
+        }
+    }
+
+    pub fn pre_entrypoint(
+        plugins: &mut Vec<Box<dyn Plugin>>,
+        gas_meter: &mut impl GasMeter,
```

Comment:
can the `GasMeter` be confined into the implementation of the `Plugin`?

Reply:
I think so! Will explore further in the gas refactor